// @sap/cds-dk/lib/init/util/spinner.js

const SPINNER_THEMES = {
    spin: ['—', '\\', '|', '/'],
    bounce: ['_', '—', '‾', '—', '/', '|', '/', '—'],
    nightRider: ['o...', '.o..', '..o.', '...o', '..o.', '.o..'],
    balloon: ['o', 'O']
}

const RENDER_DELAY = 1250;

const ESC_SEQ = '\u001B[';
const HIDE_CURSOR = ESC_SEQ + '?25l';
const SHOW_CURSOR = ESC_SEQ + '?25h';
const MOVE_CURSOR_UP = ESC_SEQ + '1A';
const ERASE_LINE = ESC_SEQ + 'K';
const MOVE_CURSOR_START = '\r';

class Spinner {
    constructor(logger, theme = Spinner.THEME_SPIN, initialDelay = 1000) {
        this.logger = logger;
        this.theme = theme;
        this.initialDelay = initialDelay;
    }

    static get THEME_RANDOM() {
        const keys = Object.keys(SPINNER_THEMES);
        const idx = Math.round(Math.random() * (keys.length - 1));
        return keys[idx];
    }

    async run(msg, cb, postfix = '') {
        try {
            this._start(msg);
            // use return await to catch errors here
            return await cb();
        } finally {
            this._stop(msg, postfix);
        }
    }

    _start(msg) {
        if (!process.env.DEBUG && this.logger.isTTY) {
            this.logger.log(msg + HIDE_CURSOR + MOVE_CURSOR_UP);
            this.timeoutId = setTimeout(() => {
                let idx = 0;
                const prefix = MOVE_CURSOR_START + ERASE_LINE + msg + ' ';
                this.intervalId = setInterval(() => {
                    this.logger.log(prefix + SPINNER_THEMES[this.theme][idx] + MOVE_CURSOR_UP);
                    idx = (idx + 1) % SPINNER_THEMES[this.theme].length;
                }, RENDER_DELAY);
            }, this.initialDelay);
        } else {
            this.logger.log(msg);
        }
    }

    _stop(msg, postfix) {
        if (!process.env.DEBUG && this.logger.isTTY) {
            if (this.timeoutId) {
                clearTimeout(this.timeoutId);
                this.timeoutId = null;
            }

            if (this.intervalId) {
                clearInterval(this.intervalId);
                this.intervalId = null;
            }

            this.logger.log(MOVE_CURSOR_START + ERASE_LINE + msg + ' ' + postfix + SHOW_CURSOR);
        }
    }
}

Spinner.THEME_SPIN = 'spin';
Spinner.THEME_BOUNCE = 'bounce';
Spinner.THEME_NIGHT_RIDER = 'nightRider';
Spinner.THEME_BALLOON = 'balloon';

module.exports = Spinner;
